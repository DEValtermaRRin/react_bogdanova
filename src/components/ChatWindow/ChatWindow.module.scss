.chat {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 560px;
  max-height: 99%;
  background-color: #161414;
  border-radius: 1rem;
  display: flex;
  justify-content: space-between;
  padding: 5px 5px;
  &__light{
    background-color: #ccc;
  }
}

.userform {
  display: flex;
  justify-content: space-between;
}


