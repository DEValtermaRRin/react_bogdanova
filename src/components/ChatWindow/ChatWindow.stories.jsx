import React from 'react';
import { Chat } from './Chat';

export default {
  component: Chat,
};

const Template = (args) => <Chat {...args} />;

export const Test = Template.bind({});
