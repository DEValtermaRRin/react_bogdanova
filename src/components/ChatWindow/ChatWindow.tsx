import React, { FC, useState } from 'react';
import { useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';

import { MessageList } from './components/MessageList';
import { User } from './components/User';
import { Form } from './components/Form';
import { ShowUser } from './components/ShowUser';

import { selectChatList } from 'src/store/chatlist/selectors';
import { WithClasses } from 'src/HOC/WithClasses';

import style from './ChatWindow.module.scss';

export const ChatWindow: FC = () => {
  const { chatId } = useParams();
  const [userName, setUserName] = useState('');
  const MessageListWithClass = WithClasses(MessageList);

  const chatList = useSelector(
    selectChatList,
    // (prev, next) => prev.length === next.length,
  );

  return (
    <div className={style.chat}>
      <div className={style.userform}>
        <User name={userName} getName={setUserName} />
        <ShowUser userName={userName} />
      </div>
      <MessageListWithClass
        messages={chatId ? chatList[chatId] : []}
        classes={style.messagelist}
      />
      <Form userName={userName} />
    </div>
  );
};
