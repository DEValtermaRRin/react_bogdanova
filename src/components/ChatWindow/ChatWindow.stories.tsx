import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { ChatWindow } from './ChatWindow';
import style from '.ChatWindow.module.scss';

export default {
  title: 'ChatWindow',
  component: ChatWindow,
  argTypes: {
    variant: {
      type: 'string',
      description: 'Окно чата',
      defaultValue: style.chat,
      options: [style.chat__light],
      control: { type: 'color' },
    },
  },
} as ComponentMeta<typeof ChatWindow>;

const Template: ComponentStory<typeof ChatWindow> = () => <ChatWindow />;

export const Default = Template.bind({});

Default.arguments = { variant: style.chat };
