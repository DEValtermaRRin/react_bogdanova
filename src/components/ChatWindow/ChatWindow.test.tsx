import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import userEvent from '@testing-library/user-event';
import { ChatWindow } from './ChatWindow';

describe('ChatWindow', () => {
  it('render component', () => {
    const mockFn = jest.fn();
    render(<ChatWindow messages={{ a: [] }} setMessages={mockFn} />);
  });
  it('render with snapshot', () => {
    const mockFn = jest.fn();
    const { asFragment } = render(
      <ChatWindow messages={{}} setMessages={mockFn} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('field message is in ChatWindow', () => {
    const mockFn = jest.fn();
    render(<ChatWindow messages={{}} setMessages={mockFn} />);
    expect(screen.getByTestId('button')).toBeInTheDocument();
  });
  it('message output check', () => {
    const mockFn = jest.fn();
    render(<ChatWindow messages={{}} setMessages={mockFn} />);
    fireEvent.input(screen.getByTestId('message'), {
      target: { value: 'Some message' },
    });
    fireEvent.click(screen.getByTestId('button'));
    expect(screen.getByTestId('message')).toContainHTML('Some message');
  });
  it('answer from bot', async () => {
    const mockFn = jest.fn();
    render(<ChatWindow messages={{}} setMessages={mockFn} />);
    fireEvent.input(screen.getByTestId('user'), {
      target: { value: 'Vasya' },
    });
    await userEvent.click(screen.getByTestId('button'));
    expect(
      setTimeout(
        () => screen.findByText('BOT: You entered an empty message'),
        1100,
      ),
    ).toBeTruthy();
  });
  it('answer bot on empty message', () => {
    const mockFn = jest.fn();
    const { asFragment } = render(
      <ChatWindow messages={{}} setMessages={mockFn} />,
    );
    fireEvent.input(screen.getByTestId('message'), {
      target: { value: '' },
    });
    fireEvent.click(screen.getByTestId('button'));
    expect(asFragment()).toMatchSnapshot();
  });
});
