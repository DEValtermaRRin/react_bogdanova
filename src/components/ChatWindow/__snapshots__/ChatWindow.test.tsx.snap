// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatWindow answer bot on empty message 1`] = `
<DocumentFragment>
  <div
    class="chat"
  >
    <div
      class="userform"
    >
      <div
        class="user"
      >
        <div
          class="user__icon"
        />
        <input
          class="user__input"
          data-testid="user"
          placeholder="Enter your name"
          type="text"
          value=""
        />
      </div>
      <p
        class="showuser"
      >
        Hello, 
      </p>
    </div>
    <div
      class="messagelist"
      data-testid="messageList"
    />
    <form
      class="form"
      data-testid="form"
    >
      <button
        class="del-button"
        type="button"
      >
        <svg
          viewBox="0 0 96 96"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <path
            d="m24,78c0,4.968 4.029,9 9,9h30c4.968,0 9-4.032 9-9l6-48h-60l6,48zm33-39h6v39h-6v-39zm-12,0h6v39h-6v-39zm-12,0h6v39h-6v-39zm43.5-21h-19.5c0,0-1.344-6-3-6h-12c-1.659,0-3,6-3,6h-19.5c-2.487,0-4.5,2.013-4.5,4.5s0,4.5 0,4.5h66c0,0 0-2.013 0-4.5s-2.016-4.5-4.5-4.5z"
          />
        </svg>
      </button>
      <textarea
        class="message"
        data-testid="message"
        placeholder="Your message"
      />
      <button
        class="btn"
        data-testid="button"
        disabled=""
        type="submit"
      >
        <svg
          class="ionicon"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Enter your name
          </title>
          <path
            d="M470.3 271.15L43.16 447.31a7.83 7.83 0 01-11.16-7V327a8 8 0 016.51-7.86l247.62-47c17.36-3.29 17.36-28.15 0-31.44l-247.63-47a8 8 0 01-6.5-7.85V72.59c0-5.74 5.88-10.26 11.16-8L470.3 241.76a16 16 0 010 29.39z"
          />
        </svg>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`ChatWindow render with snapshot 1`] = `
<DocumentFragment>
  <div
    class="chat"
  >
    <div
      class="userform"
    >
      <div
        class="user"
      >
        <div
          class="user__icon"
        />
        <input
          class="user__input"
          data-testid="user"
          placeholder="Enter your name"
          type="text"
          value=""
        />
      </div>
      <p
        class="showuser"
      >
        Hello, 
      </p>
    </div>
    <div
      class="messagelist"
      data-testid="messageList"
    />
    <form
      class="form"
      data-testid="form"
    >
      <button
        class="del-button"
        type="button"
      >
        <svg
          viewBox="0 0 96 96"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <path
            d="m24,78c0,4.968 4.029,9 9,9h30c4.968,0 9-4.032 9-9l6-48h-60l6,48zm33-39h6v39h-6v-39zm-12,0h6v39h-6v-39zm-12,0h6v39h-6v-39zm43.5-21h-19.5c0,0-1.344-6-3-6h-12c-1.659,0-3,6-3,6h-19.5c-2.487,0-4.5,2.013-4.5,4.5s0,4.5 0,4.5h66c0,0 0-2.013 0-4.5s-2.016-4.5-4.5-4.5z"
          />
        </svg>
      </button>
      <textarea
        class="message"
        data-testid="message"
        placeholder="Your message"
      />
      <button
        class="btn"
        data-testid="button"
        disabled=""
        type="submit"
      >
        <svg
          class="ionicon"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Enter your name
          </title>
          <path
            d="M470.3 271.15L43.16 447.31a7.83 7.83 0 01-11.16-7V327a8 8 0 016.51-7.86l247.62-47c17.36-3.29 17.36-28.15 0-31.44l-247.63-47a8 8 0 01-6.5-7.85V72.59c0-5.74 5.88-10.26 11.16-8L470.3 241.76a16 16 0 010 29.39z"
          />
        </svg>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
